### PR TITLE
tls: add aggregated apiserver certs.

### PR DESCRIPTION
**Pull request by diegs**
This adds a new CA and certificate pair to support running aggregated
api-servers in the cluster. This is required for deploying
metrics-server (heapster replacement) and other aggregated api-server
components that will start to appear in 1.9+.

In the self-signed scenario these are derived from the kube_ca, which
happens to also be the apiserver serving ca. In practice the kube_ca
should be the parent of both this CA and the one that is used by the
--tls-ca-file flag in the apiserver. I added a TODO to fix that when
we implement the "single CA is the parent of all generated certs" change
for track 2.

### CODE REVIEW

**Comment by diegs** [comment]
cc @brancz

File: contrib/user-provided-certs/kube/self-signed/api_proxy.tf [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+resource "tls_cert_request" "apiserver_proxy" {
+  key_algorithm   = "${tls_private_key.apiserver_proxy.algorithm}"
+  private_key_pem = "${tls_private_key.apiserver_proxy.private_key_pem}"
+
+  subject {
+    common_name  = "kube-apiserver-proxy"
+    organization = "kube-master"
+  }
+
+  dns_names = [
```
Comment (ericchiang) [inline contrib/user-provided-certs/kube/self-signed/api_proxy.tf]: I don't think this should be used as a serving cert, just a client cert.

File: contrib/user-provided-certs/kube/self-signed/api_proxy.tf [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  algorithm = "RSA"
+  rsa_bits  = "2048"
+}
+
+resource "tls_cert_request" "apiserver_proxy" {
+  key_algorithm   = "${tls_private_key.apiserver_proxy.algorithm}"
+  private_key_pem = "${tls_private_key.apiserver_proxy.private_key_pem}"
+
+  subject {
+    common_name  = "kube-apiserver-proxy"
```
Comment (ericchiang) [inline contrib/user-provided-certs/kube/self-signed/api_proxy.tf]: Common name has to match the value specified by:

```
--requestheader-allowed-names
```

Comment (diegs) [reply contrib/user-provided-certs/kube/self-signed/api_proxy.tf]: Ah, I thought that was the CN in the CA. The docs are very unclear, and the metrics server appeared to be working. But I'll change the flag to this one in the operators-PR.

File: contrib/user-provided-certs/kube/self-signed/api_proxy.tf (hunk 1 above)
Comment (diegs) [reply contrib/user-provided-certs/kube/self-signed/api_proxy.tf]: So, remove the dns names and ip addresses?

**Comment by diegs** [comment]
@cpanato @enxebre any hints on how to debug the failed custom_tls tests? Do I need to generate the new certs somehow? The `user_provided_tls` stuff in the smoke test directory doesn't have much documentation.

File: contrib/user-provided-certs/kube/self-signed/api_proxy.tf (hunk 1 above)
Comment (ericchiang) [reply contrib/user-provided-certs/kube/self-signed/api_proxy.tf]: Yes. Allowed usages bellow should also drop "server_auth"

Comment (diegs) [reply contrib/user-provided-certs/kube/self-signed/api_proxy.tf]: Done

**Comment by diegs** [comment]
Ah, think I found the change I needed to make for smoke tests.

**Comment by diegs** [comment]
zomg ruby linter doesn't like methods with more than 30 lines?

**Comment by diegs** [comment]
I think azure flaked.

**Comment by brancz** [comment]
On a high level this looks good to me. I'm wondering how aggregated APIs are going to get their server certs, seems like we want something like the certificates API, but for multi CA (kubernetes components, etcd, aggregated APIs,...). Do we have a solid idea how to solve that?

**Comment by ericchiang** [comment]
> On a high level this looks good to me. I'm wondering how aggregated APIs are going to get their server certs, seems like we want something like the certificates API, but for multi CA (kubernetes components, etcd, aggregated APIs,...). Do we have a solid idea how to solve that?

Aggregated API servers get to advertise their own CA, which the API server then trusts https://github.com/kubernetes/kubernetes/blob/v1.9.2/staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go#L56

So the metrics server can generate a self-signed one and that should be fine.

**Comment by brancz** [comment]
lgtm

**Comment by diegs** [comment]
@brancz @ericchiang plot twist: apparently we *don't* want the metrics server to generate self-signed certs?

https://github.com/kubernetes-incubator/metrics-server/issues/37#issuecomment-361686213

**Comment by diegs** [comment]
retest this please. azure and bare metal.

**Comment by diegs** [comment]
All tests passed!

**Comment by diegs** [comment]
May I please get a review on the installer side? @enxebre? @squat?

**Comment by ericchiang** [comment]
We probably want to provide the self-signed certs through a secret.

Mentioned this on JIRA, but I'm going to follow up on https://github.com/kubernetes/kubernetes/issues/54960 to see if there are any specific concerns around this strategy. I'm not familiar with the subtitles https://github.com/kubernetes/kubernetes/issues/54960

File: platforms/aws/tectonic.tf [hunk 3]
```diff
@@ -28,18 +28,21 @@ module "bootkube" {
...
-  etcd_peer_key_pem    = "${module.etcd_certs.etcd_peer_key_pem}"
-  etcd_server_cert_pem = "${module.etcd_certs.etcd_server_crt_pem}"
-  etcd_server_key_pem  = "${module.etcd_certs.etcd_server_key_pem}"
-  kube_ca_cert_pem     = "${module.kube_certs.ca_cert_pem}"
-  kubelet_cert_pem     = "${module.kube_certs.kubelet_cert_pem}"
-  kubelet_key_pem      = "${module.kube_certs.kubelet_key_pem}"
+  aggregator_ca_cert_pem   = "${module.kube_certs.aggregator_ca_cert_pem}"
+  apiserver_cert_pem       = "${module.kube_certs.apiserver_cert_pem}"
+  apiserver_key_pem        = "${module.kube_certs.apiserver_key_pem}"
+  apiserver_proxy_cert_pem = "${module.kube_certs.apiserver_cert_pem}"
```
Comment (enxebre) [inline platforms/aws/tectonic.tf]: wouldn't this need to be `module.kube_certs.apiserver_proxy_cert_pem` and `module.kube_certs.apiserver_proxy_key_pem`? same question for all platforms

File: modules/tls/kube/self-signed/aggregator_ca.tf [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# Kubernetes Aggregated API Server CA (resources/generated/tls/aggregator-ca.crt)
+#
+# TODO(diegs): this should be a sibling of the `--tls-ca-file` CA. However the
```
Comment (enxebre) [inline modules/tls/kube/self-signed/aggregator_ca.tf]: shall we add same todo in contrib module?

**Comment by enxebre** [comment]
Hey @diegs overall looks good to me but If I'm not missing anything it seems we are generating the proxy certs but those are consumed nowhere?

File: modules/tls/kube/self-signed/aggregator_ca.tf (hunk 4 above)
Comment (diegs) [reply modules/tls/kube/self-signed/aggregator_ca.tf]: Done

File: platforms/aws/tectonic.tf (hunk 3 above)
Comment (diegs) [reply platforms/aws/tectonic.tf]: Great catch, thank you! Changed to the correct value.

**Comment by diegs** [comment]
retest this please. azure and bare metal. bare metal worked great on my machine.

**Comment by diegs** [comment]
retest this please. the old failures passed, but ones that passed failed :-/

**Comment by diegs** [comment]
retest this please. i don't know why they keep failing. logs show that the cluster just doesn't come up completely. @cpanato perhaps you have more insights? thanks!

**Comment by cpanato** [comment]
@diegs for the metal tests it failed because the certificates we use in the tests expired. I already fixed that, will retest this.

**Comment by cpanato** [comment]
@diegs can you please rebase?

**Comment by diegs** [comment]
@cpanato rebased.

**Comment by diegs** [comment]
Azure failure:

```

Error: Error applying plan:

1 error(s) occurred:

* module.vnet.azurerm_virtual_network.tectonic_vnet: 1 error(s) occurred:

* azurerm_virtual_network.tectonic_vnet: network.VirtualNetworksClient#CreateOrUpdate: Failure sending request: StatusCode=200 -- Original Error: Long running operation terminated with status 'Failed': Code="InternalServerError" Message="An error occurred."

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.


make: *** [apply] Error 1
Couldn't find output "api_ip_addresses" for module var: module.vnet.api_ip_addresses
```

**Comment by cpanato** [comment]
@diegs this is an azure issue. I have a ticket to we investigate that.

As soon this build finish I will re-trigger because aws/govcloud was missing in the current test build

**Comment by cpanato** [comment]
retest this please.

**Comment by diegs** [comment]
retest this please. govcloud and azure.

**Comment by diegs** [comment]
govcloud:

```
Error: Error applying plan:

1 error(s) occurred:

* aws_vpn_gateway.vpg (destroy): 1 error(s) occurred:

* aws_vpn_gateway.vpg: IncorrectState: The VPN gateway is in use.
	status code: 400, request id: 1d88c56e-d27f-4c6f-a01e-7f6e83624ab8
```

azure:

some pods never came up.

**Comment by diegs** [comment]
retest this please.

**Comment by diegs** [comment]
I give up.

**Comment by cpanato** [comment]
all green now @diegs

File: modules/tls/kube/self-signed/outputs.tf [hunk 5]
```diff
@@ -1,3 +1,11 @@
+output "aggregator_ca_cert_pem" {
+  value = "${tls_locally_signed_cert.aggregator_ca.cert_pem}"
+}
+
+output "aggregator_ca_key_pem" {
```
Comment (squat) [inline modules/tls/kube/self-signed/outputs.tf]: I'm curious: why does this module output an aggregator_ca_key but not the tls/kube/user-provided?

It seems like this output is not consumed anywhere either.

File: contrib/user-provided-certs/kube/self-signed/aggregator_ca.tf [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  ca_private_key_pem    = "${tls_private_key.kube_ca.private_key_pem}"
+  ca_cert_pem           = "${tls_self_signed_cert.kube_ca.cert_pem}"
+  validity_period_hours = "${var.validity_period}"
+
+  allowed_uses = [
+    "key_encipherment",
+    "digital_signature",
+    "client_auth",
+  ]
+}
```
Comment (squat) [inline contrib/user-provided-certs/kube/self-signed/aggregator_ca.tf]: Should this module also create local files for the aggregated ca key and crt just like `modules/tls/kube/self-signed` does? i.e.:
```hcl
resource "local_file" "aggregator_ca_key" {
...
}
resource "local_file" "aggregator_ca_crt" {
...
}
```

File: modules/tls/kube/self-signed/outputs.tf (hunk 5 above)
Comment (diegs) [reply modules/tls/kube/self-signed/outputs.tf]: You're right, it's not a required output. Got copy-paste happy. Removed.

File: contrib/user-provided-certs/kube/self-signed/aggregator_ca.tf (hunk 6 above)
Comment (diegs) [reply contrib/user-provided-certs/kube/self-signed/aggregator_ca.tf]: Done

**Comment by squat** [comment]
ok to test

**Comment by squat** [comment]
dang, we have some merge conflicts. Conflicts are only on markdown files. All tests have passed for TF. I'm removing the labels and pushing an extra commit to fix the conflicts with the docs.

**Comment by cpanato** [comment]
ok

**Comment by squat** [comment]
@diegs: the tls module docs you wrote have been PRd to https://github.com/coreos/tectonic-docs/pull/135 now that they no longer exist in this repo since https://github.com/coreos/tectonic-installer/pull/2678

**Comment by diegs** [comment]
@squat thanks!

**Pull request merged**